ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+
+Given these limitations in Kubernetes, it becomes clear that we can benefit with a mechanism for abstracting workspaces from their underlying namespaces. The approach proposed here to best represent the workspace concept is to extend the kubernetes API via a new CRD. The Konflux user will be able to interact with workspaces as a resource of kubernetes. They will be able to execute all CRUD operations to manage their workspace lifecycle.
+
+### Architecture
+
+The workspace team has factored these limitations in while developing features on Konflux downstream by creating a layer between the Konflux User Interfaces and the KubeSaw components that provide namespace as a service (NSaaS) layer. We understand that there are multiple ways to implement a namespace provisioning service depending on the constraints of the system. Our first implementation is KubeSaw, but as we redefined some constraints on Konflux deployments, other backends for a workspace became viable. See the diagram below for the different backends that can be implemented at the moment.
+
+#### Diagram
+
+![](../diagrams/ADR-0040/konflux-workspaces.png)
+
+#### Description
+
+The diagram depicts 3 backend implementations for the Workspace Controller: KubeSaw, Local and GitOps. All backends end up managing namespaces, their initial resources required, and connectivity between these namespaces to authenticated identities.
+
+##### KubeSaw
+
+This is the current implementation downstream running today in our KubeSaw deployments. It leverages the KubeSaw host operator, member operator and registration service components. It is designed to scale with multiple member cluster support. This implementation will only be possible with OpenShift clusters, since there is no vanilla support of K8S at the moment. Implementing support for non-OpenShift Kubernetes deployments is scheduled for the future.
+
+##### Local
+
+This implementation assumes Konflux runs in single cluster mode. It is a straightforward implementation of local namespace provisioning, bootstrapping and user allocation. It will require a workspace config object that describes the default resources to set during the bootstrapping phase.
+
+##### GitOps
+
+This implementation provisions the namespace in a target cluster, but the bootstrapping is delegated to a GitOps provider like ArgoCD. A new application will be spawned on the git repo and ArgoCD will track those resources and synchronize them with the cluster. This is the current approach using upstream at the moment.
```

Comment:
1. can you please provide details about how this will be implemented? which changes would be needed to https://github.com/konflux-workspaces/workspaces ?

2. As I mentioned above, upstream Konflux isn't opinionated about the way namespace are provisioned with initial set of resources. We don't need to create any resource in a namespace for getting Konflux to operate on it.

Reply:
We've architected konflux-workspaces so that our storage layer is swappable, and it could be configured with a different backend (see `server/persistence`).  Right now, we've only implemented a backend that can talk to kubesaw, but there's no reason we can't implement more there besides developer throughput.

We'd also need to plumb through this configuration during server startup (such as read the backend choice from a configmap and set it up appropriately).  As always, we'd also need to implement tests for this.

Reply:
@sadlerap which interfaces should be implemented in order to add a new provider?

Reply:
We'd need to implement the interfaces represented in the [`ReadClient`](https://github.com/konflux-workspaces/workspaces/blob/main/server/persistence/readclient/readclient.go#L15-L20) and [`WriteClient`](https://github.com/konflux-workspaces/workspaces/blob/main/server/persistence/writeclient/writeclient.go#L18-L24) structs.  Of particular interest would be the [`DirectAccessChecker`](https://github.com/konflux-workspaces/workspaces/blob/main/server/persistence/clientinterface/readclient.go#L46-L48) interface, as this is how we check if a user has access to a workspace.

On further review, we'd also need to make changes in the `operator/internal/controller` package.  Currently, the `InternalWorkspace` CRD gets converted to a `Space` that manages the namespace controlling that workspace, along with granting users correct permissions via invoking `SpaceBindings`.  If we were to use a separate backend not based on kubesaw, we'd need to swap out these resources for their equivalent semantics.